### ISSUE TITLE
mask_reasoning_content does not mask template-inserted empty

### ISSUE THREAD

**Issue by qwerty3564** [NVIDIA-NeMo/Automodel#3738]
# `mask_reasoning_content` does not mask template-inserted empty `<think></think>`

## Problem

When fine-tuning Nemotron 3.5 with:

```yaml
mask_reasoning_content: true
```

the empty `<think></think>` automatically inserted by the chat template is still included in the loss if the dataset has no explicit `reasoning_content`.

Example dataset:

```json
{
  "messages": [
    {
      "role": "user",
      "content": "Solve the following programming problem..."
    },
    {
      "role": "assistant",
      "content": "import sys\n..."
    }
  ]
}
```

After applying the Nemotron chat template, it becomes roughly:

```text
<|im_start|>assistant
<think></think>import sys...
<|im_end|>
```

However, `mask_reasoning_content` only triggers when the original assistant message contains a non-empty `reasoning_content` field.

Therefore, for code-only SFT, the effective loss is roughly:

```text
<think>      -> LOSS
</think>     -> LOSS
code         -> LOSS
```

## Why this may be a problem

With thousands of non-thinking SFT samples, training repeatedly reinforces:

```text
<think> -> </think>
```

But when `enable_thinking=True`, the inference prompt ends with:

```text
<think>
```

and the model is expected to generate reasoning before producing `</think>`.

Repeatedly supervising the immediate `<think> -> </think>` transition during non-thinking SFT may bias the fine-tuned model toward terminating reasoning immediately and potentially degrade its original thinking capability.

## Expected behavior

It would be useful to support an option that keeps the empty `<think></think>` in `input_ids`, but masks those tokens from the loss:

```text
input:
question <think> </think> code

labels:
-100    -100   -100     code
```

For example:

```yaml
mask_empty_reasoning_tags: true
```

This would preserve the Nemotron non-thinking prompt format while avoiding unnecessary supervision of immediate reasoning termination.

**Comment by HuiyingLi** [comment]
Thank you @qwerty3564 for the issue!
@khazic could you please take a look, appreciate it!

**Comment by khazic** [comment]
Thanks for the detailed report. I reproduced this: with a chat template that inserts an empty `<think></think>` for assistant turns that carry no `reasoning_content`, both tag tokens stay in the loss even with `mask_reasoning_content: true`. The current masking works by re-rendering the conversation with `reasoning_content` cleared and diffing the two token sequences, so an empty field produces no diff and nothing gets masked.

I will open a PR for this. Since the same pattern (template-inserted empty thinking tags in non-thinking turns) shows up in several other chat templates, not only Nemotron, the fix will target that case generically rather than special-casing Nemotron.

**Comment by khazic** [comment]
Opened #3739 for this.

The option is called `mask_generation_prompt` rather than `mask_empty_reasoning_tags`, because it detects the span from the template instead of from tag strings: for each assistant turn it renders the generation prompt and excludes the tokens that prompt supplies (the role header plus any inserted reasoning block). That covers your Nemotron case and the same pattern in Qwen3, GLM, DeepSeek-V3.1 and others, and it stays correct for a template like Kimi-K2-Thinking, which inserts the same empty block but does not supply it in the generation prompt, so the model does generate it and it should stay in the loss.

`input_ids` are unchanged, the option is off by default, and the PR lists the chat templates it was verified against.
